### ISSUE TITLE
Text Selection

### ISSUE THREAD

**Issue by shanehandley** [hecrj/iced#36]
It would be nice to be able to select a range of text with the mouse/+keyboard. I think this would be required for copying sections of text when clipboard support is added.

**Comment by remingtonc** [comment]
This is critically important for productivity-related applications. Copy/paste of text content is fundamental to many user workflows. The current workaround would be to expose the content in a `text_input` widget?

**Comment by krooq** [comment]
As mentioned this is already supported for `TextInput` perhaps the code can be reused.

**Comment by Atreyagaurav** [comment]
Adding on to this.

For text, as well as for Markdown, not being able to select to copy things means the information displayed is only for visual purposes. The only workaround for selection purpose seems to be using the text_editor, but that does not match the aesthetics or the complication required for the task.

**Comment by samuele-ruffino96** [comment]
Actually you could obtain **text selection** capabilities using the existing `text_editor` widget. 

Simply handle the messages but ignore the actual editing operations: `Action::Edit(_)`

```rust
use iced::widget::{container, text_editor};
use iced::{Element, Fill, Task, Theme};

pub fn main() -> iced::Result {
    iced::application(SimpleTextViewer::new, SimpleTextViewer::update, SimpleTextViewer::view)
        .theme(|_| Theme::Dark)
        .run()
}

struct SimpleTextViewer {
    content: text_editor::Content,
}

#[derive(Debug, Clone)]
enum Message {
    Edit(text_editor::Action),
}

impl SimpleTextViewer {
    fn new() -> (Self, Task<Message>) {
        let content = text_editor::Content::with_text(
            "This is a selectable read-only text viewer\n\nYou can:\n- Select text with mouse\n- Copy text with Cmd+C\n- Select all text with Cmd+A\n- Navigate with arrow keys\n- Scroll through content\n- Select across multiple lines\n\nBut you cannot edit the content"
        );

        (Self { content }, Task::none())
    }

    fn update(&mut self, message: Message) -> Task<Message> {
        match message {
            Message::Edit(action) => {
                // Only block actual editing actions
                match action {
                    // Block editing actions (these contain the actual editing operations)
                    text_editor::Action::Edit(_) => {
                        // Do nothing - ignore all editing operations
                    }
                    // Allow all other actions (movement, selection, clicking, scrolling, etc.)
                    _ => {
                        self.content.perform(action);
                    }
                }
            }
        }
        Task::none()
    }

    fn view(&self) -> Element<'_, Message> {
        container(
            text_editor(&self.content)
                .placeholder("No content to display...")
                .on_action(Message::Edit)
        )
            .width(Fill)
            .height(Fill)
            .padding(20)
            .into()
    }
}
```
Live demonstration: https://github.com/user-attachments/assets/e95511ec-86b6-4117-b4e7-fde384a3fb36

**Comment by Atreyagaurav** [comment]
That only works for something with just text. I do use it when I want that like logs and such. But for text with spans like rich text/markdown can't be inserted in text_editor right? Unless there's a way to do that (maybe through highlights). But writing a custom highlights has a bug that it doesn't always trigger while insertion of huge text chunk (#2946 )

Edit: I just noticed that the comments (mine included) already mention using text_editor being a workaround. But it's not enough.

**Comment by pml68** [comment]
I'm working on selectable rich text for Foghorn which includes support for keyboard selection shortcuts. For some reason, all glyphs in the `Paragraph` have a reported y (and y_offset) of 0, so I can't correctly draw the [selection rectangle(s)](https://github.com/foghorn-rs/foghorn/blob/main/widget/src/text/rich.rs#L406-L412), but selection logic works. I'm not sure if this is a bug in cosmic-text or iced's `Paragraph` setup.

If you don't need keyboard support you could check out [Halloy's implementation](https://github.com/squidowl/halloy/blob/main/src/widget/selectable_rich_text.rs)

**Comment by NicTanghe** [comment]
This is the first problem i hit. 2d ui that doesn't let users select UI is broken.

Has annyone done anny work on this and can i create a pull request ?

**Comment by pml68** [comment]
It's on the roadmap for iced, I've got https://docs.rs/iced_selection, see the to-do section.

**Comment by NicTanghe** [comment]
for me ideally this is a global toggle and not a special new type of paragraph only.

**Comment by airstrike** [comment]
a global toggle doesn't quite make sense as not all UI text is meant to be selectable. you can't select the text in your browser's tab title, for instance.

a `.selectable()` builder method feels like the right API imho

**Comment by NicTanghe** [comment]
I want to be able to select all the text in my UI,
That doesn't mean I will.

I whould definately want to select text in browser tab titles and,
being able to select the text in the tab titles isn't going to prevent me from not selecting it.

But not being able to select it will prevent me to select it.
it's a bit of a no-brainer to me.

On the web all the text is selectable.
So it does make sense.
And people generally get really upset when they can't select text in their UI.

It's also way less work to just create a global toggle. That is on by default.

People are civil here becouse they aren't customers that are complaining that they still can't select text in their UI but i've been on the phone with plenty of people wanting to cast the programmers of their dayly use app into hell.

This really isn't a decision that programmers should make for users. Just default all to on.

**Comment by airstrike** [comment]
> I want to be able to select all the text in my UI, That doesn't mean I will.
> 
> I whould definately want to select text in browser tab titles and, being able to select the text in the tab titles isn't going to prevent me from not selecting it.
> 
> But not being able to select it will prevent me to select it. it's a bit of a no-brainer to me.
> 
> On the web all the text is selectable. So it does make sense. And people generally get really upset when they can't select text in their UI.
> 
> It's also way less work to just create a global toggle. That is on by default.
> 
> People are civil here becouse they aren't customers that are complaining that they still can't select text in their UI but i've been on the phone with plenty of people wanting to cast the programmers of their dayly use app into hell.
> 
> This really isn't a decision that programmers should make for users. Just default all to on.

I'm not convinced. Prior art suggests otherwise. If you want, you could still make your entire text `.selectable()` in your app.

Your suggested approach forces me to make my entire app selectable.

My suggested approach lets you make your entire app selectable while preserving my choice to make a sane UI that is selectable only where it matters.

Therefore, my suggestion is a strictly superior paradigm. QED.

**Comment by edwloef** [comment]
It could also make use of a feature flag similar to `crisp` that just changes the default, and a `.selectable(bool)` or something to opt out of the default? Seems to me like the best of both worlds.

**Comment by pml68** [comment]
That's probably the best solution, also similarly to text shaping

**Comment by Atreyagaurav** [comment]
I don't know if it has been discussed already, but can I add that the global option makes sense when we also have to track how many active selections we can have.

For a normal UI, it should only have one active selection per window. But if you are using panes, it should be able to have one selection per pane because we don't want to lose the selection while moving between panes. 

So even if you have `.selectable()` for items that have text, the selection should still be tracked globally.

**Comment by pml68** [comment]
The solution @alex-ds13 proposed to this was having a wrapper where every selectable text widget within that wrapper will have one combined selection. Works well enough currently, though it's on a separate branch of iced_selection due to a few small quirks I'd like to smooth out first.

**Comment by airstrike** [comment]
> The solution [@alex-ds13](https://github.com/alex-ds13) proposed to this was having a wrapper where every selectable text widget within that wrapper will have one combined selection. Works well enough currently, though it's on a separate branch of iced_selection due to a few small quirks I'd like to smooth out first.

I agree this makes sense. Multiple selections is a widget concern, unless we lifted the concept of selections into the shell which I don't think tracks.

But making all text globally selected or opt-in via `.selectable()` is orthogonal to tracking multi-selection.

**Comment by pml68** [comment]
In the end what really matters is how @hecrj decides to do it. Not sure if he has anything planned currently (even on an abstract level), but we'll see. What I'm doing for iced_selection is considered a hack I believe. It works, but it's still a hack.

**Comment by wilsonglasser** [comment]
Hey got bitten by it in my own project (a small ssh client) and ended up just writing the thing. Sorry for not elaborating or discussing this before.
Tried to stay close to existing patterns. Opt-in, no behavior change for existing apps. There's one part i'm not proud of I needed to make `Rich::State` `pub` so a coordinator widget could drive selection across siblings. Happy to redesign that.

PR: https://github.com/iced-rs/iced/pull/3315
